solr/core/src/java/org/apache/solr/security/AllowListUrlChecker.java

```diff
@@ -154,6 +157,29 @@ public void checkAllowList(List<String> urls, ClusterState clusterState)
     }
   }
 
+  /**
+   * Gets the set of live hosts urls (host:port) built from the set of live nodes.
+   * The set is cached to be reused until the live nodes change.
+   */
+  private Set<String> getLiveHostUrls(ClusterState clusterState) {
+    if (clusterState == null) {
+      return Set.of();
+    }
+    Set<String> liveNodes = clusterState.getLiveNodes();
+    if (liveHostUrlsCache == null || liveNodes != liveNodesCache) {
+      liveHostUrlsCache = buildLiveHostUrls(liveNodes);
+      liveNodesCache = liveNodes;
+    }
```

Comment:
I'm a bit nervous of some race... maybe a synchronized is simple and in practice shouldn't bottleneck as live nodes doesn't change?